### PR TITLE
fix non-determinstic failure in higherorderfns (?)

### PR DESCRIPTION
**Pull request by SobhanMP**
Possible fix for #178. My guess is that `sprandn` is returning non-structural zeros. This patch removes those non-structural zeros.

### CODE REVIEW

**Comment by Wimmerer** [comment]
Can we just use some small known sparse matrix for this instead of using sprand until we hit one with the desired properties?

**Comment by dkarrasch** [comment]
Can't you simply `setindex!` elements to either zero or something that yields zero under function application? I mean, under the assumption that zeros and/or zero values cause the issue.

**Comment by SobhanMP** [comment]
sure, that would also work

**Comment by SobhanMP** [comment]
i think this is the best solution. the change is such that if this is not the solution, the tests will keep failing.

File: test/higherorderfns.jl [hunk 1]
```diff
@@ -8,24 +8,35 @@ module HigherOrderFnsTests
...
+    # --> test map entry point
+    fA = Array(A)
+    shapeA = size(A)
+    @test map(sin, A) == sparse(map(sin, fA))
+    @test map(cos, A) == sparse(map(cos, fA))
+    # --> test map! entry point
+    fX = copy(fA); X = similar(A)
+    map!(sin, X, A); X = similar(A) # warmup for @allocated
+    map!(sin, X, A); X = similar(A) # warmup for @allocated
+    @test alloc_tests || @allocated(map!(sin, X, A)) == 0
```
Comment (dkarrasch) [inline test/higherorderfns.jl:25]: Is this meant to be
```suggestion
    alloc_tests && @test @allocated(map!(sin, X, A)) == 0
```
?

Comment (SobhanMP) [reply test/higherorderfns.jl:25]: does it really matter?

Comment (SobhanMP) [reply test/higherorderfns.jl]: it was mean to be the other way around

**Comment by ViralBShah** [comment]
Can we update this PR to say which issue it is closing and put in a little bit description in the OP (the first comment in the PR which is empty)?

**Review by ViralBShah — APPROVED** [review]

File: test/higherorderfns.jl [hunk 2]
```diff
@@ -8,24 +8,39 @@ module HigherOrderFnsTests
 
 using Test
 using SparseArrays
+using SparseArrays: nonzeros
```
Comment (dkarrasch) [inline test/higherorderfns.jl:11]: Seems like `nonzeros` is already exported by `SparseArrays`.
```suggestion
```

File: test/higherorderfns.jl [hunk 3]
```diff
@@ -8,24 +8,39 @@ module HigherOrderFnsTests
...
+    shapeA = size(A)
+    @test map(sin, A) == sparse(map(sin, fA))
+    @test map(cos, A) == sparse(map(cos, fA))
+    # --> test map! entry point
+    fX = copy(fA); X = similar(A)
+    map!(sin, X, A); X = similar(A) # warmup for @allocated
+    map!(sin, X, A); X = similar(A) # warmup for @allocated
+    if @allocated(map!(sin, X, A)) == 0 || alloc_tests
+        println(stderr, "A in test_map_and_map! is ", A)
+    end
```
Comment (dkarrasch) [inline test/higherorderfns.jl:27]: I'm trying to follow the logic here, but fail. IIUC, you want this to print when `alloc_tests == false`(the third call below), which will happen when `@allocated(map!(sin, X, A)) == 0`? Do you mean
```suggestion
    if !(@allocated(map!(sin, X, A)) == 0 || alloc_tests)
        println(stderr, "A in test_map_and_map! is ", A)
    end
```
`alloc_tests = true` sounds to me like "do the allocation tests", whereas you seem to use it as "skip the allocation tests`.

**Review by dkarrasch — COMMENTED** [review]
Just some minor comments, otherwise LGTM.

**Pull request merged**